packages/core/content-manager/server/src/history/services/history.ts

```diff
@@ -395,6 +395,111 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
         pagination,
       };
     },
+
+    async restoreVersion(versionId: Data.ID) {
+      const version = await query.findOne({ where: { id: versionId } });
+      const contentTypeSchemaAttributes = strapi.getModel(version.contentType).attributes;
+      const schemaDiff = getSchemaAttributesDiff(version.schema, contentTypeSchemaAttributes);
+
+      // Set all added attribute values to null
+      const dataWithoutAddedAttributes = Object.keys(schemaDiff.added).reduce(
+        (currentData, addedKey) => {
+          currentData[addedKey] = null;
+          return currentData;
+        },
+        // Clone to avoid mutating the original version data
+        structuredClone(version.data)
+      );
+      const sanitizedSchemaAttributes = omit(
+        FIELDS_TO_IGNORE,
+        contentTypeSchemaAttributes
+      ) as Struct.SchemaAttributes;
+      // Set all deleted relation values to null
+      const dataWithoutMissingRelationsPromise = Object.entries(sanitizedSchemaAttributes).reduce(
```

Comment:
Just an idea, to make this extensible to compos, polymorphic relations, ...
Something like the `traverseEntity` utility could be used here,

That utility can deeply traverse the schema for you (dz and compos too)

you could do something like

```ts
const restoredVersion = async.pipe(
  traverseEntity(restoreRelations, ctx),
  traverseEntity(restoreMedia, ctx),
  traverseEntity(restoreComponents, ctx),
  traverseEntity(setNewAttributesToNull, ctx)
  ...
 )(sanitizedSchemaAttributes)
 ```

Not sure how compatible this example is with the traverseEntity utility, but a composition like that should be possible even with a different util here. 
And it's just an idea :) in my opinion it could make it a bit quicker to understand what parts are you restoring, and it would be easier to unit test too.

Reply:
also, you have an async.reduce util in strapi/utils ! it makes it a bit easier in cases like this if you need it

Reply:
I think I will merge as is for now for the blitz if that's ok with you. I created a ticket/subtask in the current sprint to refactor this with you in a PP session if you are up for it!